types/mapper.go

```diff
@@ -0,0 +1,232 @@
+func Register(cdc *amino.Codec) {
+	cdc.RegisterConcrete(crypto.PubKey{}, "go-crypto/PubKey", nil)
+	cdc.RegisterConcrete(crypto.PrivKey{}, "go-crypto/PrivKey", nil)
+	cdc.RegisterConcrete(crypto.Signature{}, "go-crypto/Signature", nil)
+	cdc.RegisterConcrete(sdk.StdSignature{}, "sdk/StdSignature", nil)
+	cdc.RegisterInterface((*crypto.PubKeyInner)(nil), nil)
+	cdc.RegisterConcrete(crypto.PubKeyEd25519{}, "go-crypto/PubKeyEd25519", nil)
+	cdc.RegisterConcrete(crypto.SignatureEd25519{}, "go-crypto/SignatureEd25519", nil)
+	cdc.RegisterInterface((*crypto.SignatureInner)(nil), nil)
+	cdc.RegisterInterface((*UTXOHolder)(nil), nil)
+	cdc.RegisterConcrete(BaseUTXOHolder{}, "types/BaseUTXOHolder", nil)
+}
+
+// Returns the go-Amino codec.
+func (um UTXOMapper) AminoCodec() *amino.Codec {
+	return um.cdc
+}
+
+// Returns a "sealed" utxoMapper
+// the codec is not accessible from a sealedUTXOMapper
+func (um UTXOMapper) Seal() sealedUTXOMapper {
+	return sealedUTXOMapper{um}
+}
+
+func (um UTXOMapper) GetUTXO(ctx sdk.Context, addr crypto.Address, denom uint64) UTXO {
+	store := ctx.KVStore(um.key) // Get the utxo store
+	bz := store.Get(addr) // Gets the encoded bytes at the address addr
+	// Checks to see if there is a utxo at that address
+	if bz == nil {
+		return nil
+	}
+	utxoHolder := um.decodeUTXOHolder(bz) // Decode the go-amino encoded utxoHolder
+	utxo, _ := utxoHolder.GetUTXO(denom) // Get the utxo from utxoHolder
+	return utxo
+}
+
+func (um UTXOMapper) CreateUTXO(ctx sdk.Context, utxo UTXO) {
+	addr := utxo.GetAddress() // Get the address of the utxo
+	store := ctx.KVStore(um.key) // Get the utxo store 
+	bz := store.Get(addr)
+	var utxoHolder UTXOHolder
+	if bz != nil {
+		// Holder already exists
+		utxoHolder = um.decodeUTXOHolder(bz)
+	} else {
+		// Holder does not exist and needs to be created
+		utxoHolder = NewUTXOHolder() // change
+	}
+	utxoHolder.AddUTXO(utxo) //Add the utxo to the utxoHolder
+	bz = um.encodeUTXOHolder(utxoHolder) // Encode the utxoHolder
+	store.Set(addr, bz) // Add the encoded utxo to the utxo store at address addr
+}
+
+func (um UTXOMapper) DestroyUTXO(ctx sdk.Context, utxo UTXO) {
+	addr := utxo.GetAddress()
+	store := ctx.KVStore(um.key) // Get the utxo store
+	bz := store.Get(addr)
+	if bz == nil {
+		// Add error messages
+		return
+	}
+	utxoHolder := um.decodeUTXOHolder(bz)
+	utxoHolder.DeleteUTXO(utxo)
+	if utxoHolder.GetLength() == 0 {
+		store.Delete(addr)
+	} else {
+		bz = um.encodeUTXOHolder(utxoHolder)
+		store.Set(addr, bz)
+	}
+	
+}
+
+//----------------------------------------
+// sealedUTXOMapper
+
+type sealedUTXOMapper struct {
+	UTXOMapper
+}
+
+// There's no way for external modules to mutate the 
+// sum.utxoMapper.ctx from here, even with reflection
+func (sum sealedUTXOMapper) AminoCodec() *amino.Codec {
+	panic("utxoMapper is sealed")
+}
+
+//----------------------------------------
+// misc.
+
+func (um UTXOMapper) clonePrototypePtr() interface{} {
```

Comment:
what does this do and why do we need it? (same question for `clonePrototype()` below)

Reply:
It makes a new copy of the structure that is passed into the proto field. I don't think it is necessary (it's deleted on my local branch) since we have access to the struct and can call it's constructor, but I'm not 100% sure